libraries/AC_WPNav/AC_WPNav.cpp

```diff
@@ -1066,68 +905,26 @@ bool AC_WPNav::get_vector_NEU(const Location &loc, Vector3f &vec, bool &terrain_
 }
 
-///
-/// shared methods
-///
-
-/// calc_slow_down_distance - calculates distance before waypoint that target point should begin to slow-down assuming it is travelling at full speed
-void AC_WPNav::calc_slow_down_distance(float speed_cms, float accel_cmss)
-{
-	// protect against divide by zero
-	if (accel_cmss <= 0.0f) {
-		_slow_down_dist = 0.0f;
-		return;
-	}
-    // To-Do: should we use a combination of horizontal and vertical speeds?
-    // To-Do: update this automatically when speed or acceleration is changed
-    _slow_down_dist = speed_cms * speed_cms / (4.0f*accel_cmss);
-}
-
-/// get_slow_down_speed - returns target speed of target point based on distance from the destination (in cm)
-float AC_WPNav::get_slow_down_speed(float dist_from_dest_cm, float accel_cmss)
+// helper function to calculate scurve jerk and jerk_time values
+// updates _scurve_jerk and _scurve_jerk_time
+void AC_WPNav::calc_scurve_jerk_and_jerk_time()
 {
-    // return immediately if distance is zero (or less)
-    if (dist_from_dest_cm <= 0) {
-        return WPNAV_WP_TRACK_SPEED_MIN;
-    }
-
-    // calculate desired speed near destination
-    float target_speed = safe_sqrt(dist_from_dest_cm * 4.0f * accel_cmss);
-
-    // ensure desired speed never becomes too low
-    if (target_speed < WPNAV_WP_TRACK_SPEED_MIN) {
-        return WPNAV_WP_TRACK_SPEED_MIN;
+    // calculate jerk
+    _scurve_jerk = MIN(_attitude_control.get_ang_vel_roll_max_radss() * GRAVITY_MSS, _attitude_control.get_ang_vel_pitch_max_radss() * GRAVITY_MSS);
+    if (is_zero(_scurve_jerk)) {
+        _scurve_jerk = _wp_jerk;
     } else {
-        return target_speed;
+        _scurve_jerk = MIN(_scurve_jerk, _wp_jerk);
     }
-}
 
-/// wp_speed_update - calculates how to handle speed change requests
-void AC_WPNav::wp_speed_update(float dt)
-{
-    // return if speed has not changed
-    float curr_max_speed_xy_cms = _pos_control.get_max_speed_xy();
-    if (is_equal(_wp_desired_speed_xy_cms, curr_max_speed_xy_cms)) {
-        return;
-    }
-    // calculate speed change
-    if (_wp_desired_speed_xy_cms > curr_max_speed_xy_cms) {
-        // speed up is requested so increase speed within limit set by WPNAV_ACCEL
-        curr_max_speed_xy_cms += _wp_accel_cmss * dt;
-        if (curr_max_speed_xy_cms > _wp_desired_speed_xy_cms) {
-            curr_max_speed_xy_cms = _wp_desired_speed_xy_cms;
-        }
-    } else if (_wp_desired_speed_xy_cms < curr_max_speed_xy_cms) {
-        // slow down is requested so reduce speed within limit set by WPNAV_ACCEL
-        curr_max_speed_xy_cms -= _wp_accel_cmss * dt;
-        if (curr_max_speed_xy_cms < _wp_desired_speed_xy_cms) {
-            curr_max_speed_xy_cms = _wp_desired_speed_xy_cms;
-        }
+    // calculate jerk time
+    // jounce (the rate of change of jerk) uses the attitude control input time constant because multicopters
+    // lean to accelerate meaning the change in angle is equivalent to the change in acceleration
```

Comment:
Immediate problem on heli here which we'd need to excuse.

Problem on quadplanes which are accelerating using forward thrust.

Future problem on compound helis.

Reply:
We can address this at a later date. I may also look at some formulation changes over the next year that makes this less of a problem.